### PR TITLE
compiler transform: set memory spaces

### PR DESCRIPTION
**Pull request by jorendumoulin**
This PR initializes the first pass to raise DMA calls for snax to MLIR: -set-memory-spaces

This pass makes sure that the memrefs used in the code are all assigned to a memory space.
For now this is based on simple rules:

- Memref arguments of a public function (which can be called from C code), that do not yet have an assigned memory space, are defaulted to memory space `0`.
- Linalgs can only operate on memrefs in memory space `1`.

The transformation happens in 3 steps:

1. `InitFuncMemorySpace`
In this step, all memrefs arguments of a function are assigned to a memory space. The entire function then uses this memref
2. `InitLinalgMemorySpace`
In this step, the arguments used by linalg.Generic operations are converted to be memrefs in memory space 1. Conversions are handled by `memref.memory_space_cast` operations.
3. `RealizeMemorySpaceCasts`
In this step, the memory space casts are realized by transforming them to `memref.alloc` and `memref.copy`. Some simple dependency analysis is present to determine when and if a certain memref must be copied from memory space `0` to `1` or the other way around.

### CODE REVIEW

File: tests/filecheck/transforms/set-memory-space.mlir [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// RUN: ./compiler/snax-opt %s -p set-memory-space --allow-unregistered-dialect --print-op-generic | filecheck %s
```
Comment (JosseVanDelm) [inline tests/filecheck/transforms/set-memory-space.mlir:1]: Why is --allow-unregistered-dialect necessary here, which operations are not registered??

File: compiler/transforms/set_memory_space.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: memref.MemorySpaceCast, rewriter: PatternRewriter):
+        # check for invalid memory space cast,
+        # no results => unused, removed by cse
+        if len(op.results) < 1:
+            return
+
+        # remove cast, unused op: will be removed by cse
+        if len(op.results[0].uses) == 0:
```
Comment (JosseVanDelm) [inline compiler/transforms/set_memory_space.py:160]: I don't understand this part.
Why is this necessary?
Do you mean that unused parts will be removed? Because then you are running dce (dead code elimination) instead of common subexpression elimination (CSE), no?
I believe DCE is part of `--canonicalize`, but I don't see it used in any of your tests.

File: compiler/transforms/set_memory_space.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            new_function_type,
+            region=rewriter.move_region_contents_to_new_regions(op.regions[0]),
+            visibility=op.sym_visibility,
+        )
+
+        # Replace function op
+        rewriter.replace_matched_op(new_op)
+
+
+class InitLinalgMemorySpace(RewritePattern):
```
Comment (JosseVanDelm) [inline compiler/transforms/set_memory_space.py:68]: Why do we need to insert a `memory_space_cast` (runtime cast) if we change all `memory_space_casts` in the following pass to be allocs anyway?
I feel like there is only use for this pass if afterwards some `memory_space_cast` will remain, which is not the case I believe?
Is there any `memory_space_cast` that eventually will be turned into risc-v code?

File: tests/filecheck/transforms/set-memory-space.mlir [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+"builtin.module"() ({
```
Comment (JosseVanDelm) [inline tests/filecheck/transforms/set-memory-space.mlir:3]: I feel like this test is quite complex, since it tests many things in sequence, do you think it would make sense to split up the tests?

File: compiler/transforms/set_memory_space.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+                is_input = True
+            else:
+                is_input = op.results[0] in use_op.inputs
+            if is_input:
+                # insert copy op
+                copy_op = memref.CopyOp(op.source, op.dest)
+                rewriter.insert_op_before(copy_op, use_op)
+                break
+
+        # insert copy from for first use as output
```
Comment (JosseVanDelm) [inline compiler/transforms/set_memory_space.py:198]:
```suggestion
        # insert "copy from" for first use as output
```

File: compiler/transforms/set_memory_space.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            op.results[0].type.get_shape(),
+            op.results[0].type.layout,
+            op.results[0].type.memory_space,
+        )
+
+        # Insert copy ops if newly allocated memref is used as
+        # input or output, list to visit all uses of allocated memrefs:
+        uses = [x.operation for x in op.results[0].uses]
+
+        # insert copy to for first use as input
```
Comment (JosseVanDelm) [inline compiler/transforms/set_memory_space.py:180]:
```suggestion
        # insert "copy to" for first use as input
```
Reads a bit weird otherwise

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
Cool! Please address my questions!

File: compiler/transforms/set_memory_space.py (hunk 2 above)
Comment (jorendumoulin) [reply compiler/transforms/set_memory_space.py:160]: okay, I thought DCE was part of CSE, but I could be mistaken.
These checks are mainly to make sure the pass does not apply to ops without results because the following lines would then give nasty errors:
```python
op.results[0].type.get_element_type(),
```
I'll update the comments to make this more clear

File: compiler/transforms/set_memory_space.py (hunk 3 above)
Comment (jorendumoulin) [reply compiler/transforms/set_memory_space.py:68]: Yes! Good question!

I want to stay flexible with memory casts, consider the following scenario:
There is a global address space from address 0 to FF. There are three memory spaces shown in the diagram: 5 - 6 - 8. I want them to be able to overlap each other (In snitch this case may exist for the DMA core which can access L1+L3 while an accelerator only has access to L3)

Now consider the data p1, present in memory space 8 (p1, 8)
If we want to cast it to memory space 6, this is not possible. Therefore, we must allocate + copy. However, in memory space 5 this is not necessary as this memory space already has access to p1. 

This second use case is not yet implemented, but the first conversion to memory_space_casts and only later realizing them allows for this flexibility

```
             ▲
             │xFF
       ┌──   │
       │     │
       │     │
       │8    │
┌──    │     │
│      │     ├──────  p1
│      └──   │
│            │
│5     ┌──   │
│      │     │
│      │     │
│      │     │
│      │     │
└──    │6    │
       │     │
       │     │
       │     │
       └──   │
             │
             │
             │
             │ x00
```

Comment (JosseVanDelm) [reply compiler/transforms/set_memory_space.py:68]: I think we can technically still leave them out, but okay to leave them here!

File: compiler/transforms/set_memory_space.py (hunk 2 above)
Comment (jorendumoulin) [reply compiler/transforms/set_memory_space.py]: Turns out it is indeed impossible to have a `memref.memory_space_cast` without results, so this test case is not possible. The check is therefore also not necessary, and has been removed

**Review by JosseVanDelm — APPROVED** [review]
Great!
